src/panel/Table.tsx

```diff
@@ -0,0 +1,103 @@
+    renderIconName?: (item: T) => string
+    renderGroup: (group: G) => ReactNode
+    renderCell: (column: Column<T>, itemData: T) => ReactNode
+    store: TableStore<T, G>
+}
+@observer export class Table<T, G> extends PureComponent<TableProps<T, G>> {
+    @computed get gridTemplateColumns() {
+        const {columns} = this.props;
+        return `34px ${columns.map((col, i) => `${(i === 0 ? 22 : 0) + col.width.get()}px`).join(' ')} 1fr`;
+    }
+
+    private TableItem = memo<{ isSelected: boolean, item: RowItem<T>, gridTemplateColumns: string }>(props => {
+        const { columns, store, renderIconName, renderCell } = this.props;
+        const { isSelected, item, gridTemplateColumns } = props;
+        return <div className={css('svTableRow', isSelected && 'svItemSelected')} style={{ gridTemplateColumns }}
+            ref={ele => { // TODO: ForwardRef for Group
+                if (!isSelected || !ele) return;
+                setTimeout(() => ele.scrollIntoView({ behavior: 'smooth', block: 'nearest' })); // requestAnimationFrame not working.
+            }}
+            onClick={e => {
+                e.stopPropagation();
+                store.selection.set(item);
+            }}>
+            <div></div>
+            {columns.map((col, i) => <Hi key={i} className="svTableCell"
+                style={i === columns.length - 1 ? { gridColumn: 'auto / span 2' } : {}}>
+                {i === 0 && renderIconName && <Icon name={renderIconName(item.item)} />}
+                {renderCell(col, item.item)}
+            </Hi>)}
+        </div>;
+    })
+
+    render() {
+        const {TableItem} = this;
+        const {columns, store, renderGroup, children} = this.props;
+        const {rows, selection} = store;
+        return !rows.length
+            ? children // Zero data.
+            : <div className="svTable">
+                <div className="svTableHeader" style={{ gridTemplateColumns: this.gridTemplateColumns }}>
+                    <div></div>
+                    {columns.map(col => <div key={col.name} tabIndex={0} className="svTableCell"
+                        onClick={action(() => store.toggleSort(col.name))}>
+                        {col.name}{/* No spacing */}
+                        {store.sortColumn === col.name && <Icon title="Sort" name={store.sortDir} />}
+                        <ResizeHandle size={col.width} horizontal />
+                    </div>)}
+                </div>
+                <div tabIndex={0} className={css('svTableBody', selection.get() && 'svSelected')} onKeyDown={this.onKeyDown}>
+                    {rows.map(row => {
+                        const isSelected = selection.get() === row;
+                        if (row instanceof RowGroup) {
+                            return <Hi key={row.key} className={css('svTableRow', 'svTableCell', isSelected && 'svItemSelected')}
+                                style={{ gridTemplateColumns: '6px auto auto auto auto 1fr' /* Temp */ }}
```

Comment:
>'6px auto auto auto auto 1fr' [](start = 62, length = 29)

Should these be a style somewhere that is referenced?

Reply:
Removed this temporary approach.

---
In reply to: [441691371](https://github.com/microsoft/sarif-vscode-extension/pull/282#discussion_r441691371) [](ancestors = 441691371)